packages/agents/src/index.ts

```diff
@@ -1173,17 +1193,183 @@ export async function routeAgentRequest<Env>(
+  email: ForwardableEmailMessage,
+  _env: Env
+): Promise<{ agentName: string; agentId: string } | null> {
+  const messageId = email.headers.get("message-id");
+  if (messageId) {
+    const messageIdMatch = messageId.match(/<([^@]+)@([^>]+)>/);
+    if (messageIdMatch) {
+      const [, agentId, domain] = messageIdMatch;
+      const agentName = domain.split(".")[0];
+      return { agentName, agentId };
+    }
+  }
+
+  const references = email.headers.get("references");
+  if (references) {
+    const referencesMatch = references.match(/<([A-Za-z0-9+\/]{43}=)@([^>]+)>/);
+    if (referencesMatch) {
+      const [, base64Id, domain] = referencesMatch;
+      const agentId = Buffer.from(base64Id, "base64").toString("hex");
+      const agentName = domain.split(".")[0];
+      return { agentName, agentId };
+    }
+  }
+
+  const agentName = email.headers.get("x-agent-name");
+  const agentId = email.headers.get("x-agent-id");
+  if (agentName && agentId) {
+    return { agentName, agentId };
+  }
+
+  return null;
+}
+
+export function createEmailAddressResolver<Env>(
+  defaultAgentName: string
+): EmailResolver<Env> {
+  return async (email: ForwardableEmailMessage, _env: Env) => {
+    const emailMatch = email.to.match(/^([^+@]+)(?:\+([^@]+))?@(.+)$/);
+    if (!emailMatch) {
+      return null;
+    }
+
+    const [, localPart, subAddress] = emailMatch;
+
+    if (subAddress) {
+      return {
+        agentName: localPart,
+        agentId: subAddress,
+      };
+    }
+
+    // Option 2: Use defaultAgentName namespace, localPart as agentId
+    // Common for catch-all email routing to a single EmailAgent namespace
+    return {
+      agentName: defaultAgentName,
+      agentId: localPart,
+    };
+  };
+}
+
+export function createCatchAllResolver<Env>(
+  agentName: string,
+  agentId: string
+): EmailResolver<Env> {
+  return async () => ({ agentName, agentId });
+}
+
+export type EmailRoutingOptions<Env> = AgentOptions<Env> & {
+  resolver?: EmailResolver<Env>;
+  defaultAgentName?: string;
+  defaultAgentId?: string;
+};
+
 export async function routeAgentEmail<Env>(
-  _email: ForwardableEmailMessage,
-  _env: Env,
-  _options?: AgentOptions<Env>
-): Promise<void> {}
+  email: ForwardableEmailMessage,
+  env: Env,
+  options?: EmailRoutingOptions<Env>
+): Promise<void> {
+  let routingInfo: { agentName: string; agentId: string } | null = null;
+
+  if (options?.resolver) {
+    routingInfo = await options.resolver(email, env);
+  } else {
+    routingInfo = await defaultEmailResolver(email, env);
+  }
+
+  if (!routingInfo && options?.defaultAgentName && options?.defaultAgentId) {
+    routingInfo = {
+      agentName: options.defaultAgentName,
+      agentId: options.defaultAgentId,
+    };
+  }
+
+  if (!routingInfo) {
+    console.warn("No routing information found for email, dropping message");
+    return;
+  }
+
+  const namespace = env[routingInfo.agentName as keyof Env] as AgentNamespace<
```

Comment:
I haven't tested this path yet (needs setting up workers email properly I think) - but I wonder if this is accurate.